### DISCUSSION TITLE
Custom Trace ID (from app context) for auto-instrumentation

### DISCUSSION THREAD

**Discussion by IvanVas** [open-telemetry/opentelemetry-java-instrumentation#9936] [Q&A]
Hi folks, could you help?

I'm using auto-instrumentation with java agent. The app is Jakarta EE (Jersey)
I'd like to change the Trace Id to a custom value instead of new random.
What's the best way?

I checked the extension route, but the extension code is out of main app context.

I checked the manual instrumentation, but can it be combined with auto somehow?

Cheers!

**Comment by laurit** [comment]
It is possible to replace the `IdGenerator` https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/IdGenerator.java implementation but this implementation would still need to generate ids that are valid according to https://opentelemetry.io/docs/specs/otel/trace/api/#spancontext If you get your id from an external system via http header then you should implement a custom context propagator https://opentelemetry.io/docs/specs/otel/context/api-propagators/ when you have some other reason then consider that perhaps instead of altering the trace id it might be easier to put an extra attribute `correlation.id` or whatever on the spans that are part of that trace.

**Reply by IvanVas** [reply]
Thanks for answering!

How do we replace IdGenerator for auto-instrumentation?
How to put the extra attribute, can MDC help?

**Reply by trask** [reply]
> How do we replace IdGenerator for auto-instrumentation?

you need to implement an "agent extension", see https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/extension#extensions

> How to put the extra attribute, can MDC help?

it depends, `Span.current().setAttribute()` is the first thing I'd try
